scripts/upgrade_cfme.py

```diff
@@ -0,0 +1,216 @@
+logging.basicConfig(filename=log_name, filemode="w",
+                    level=logging.DEBUG, format=formatter)
+logger = logging.getLogger("cfme_logger")
+
+console_handler = logging.StreamHandler()
+console_handler.setLevel(logging.INFO)
+console_handler.setFormatter(logging.Formatter(formatter))
+logger.addHandler(console_handler)
+
+
+class CfmeUpgradeManeger(IPAppliance):
+
+    def __init__(self, appliance_ip, repo_list, dest_version):
+        super(CfmeUpgradeManeger, self).__init__(address=appliance_ip)
+        self.repo_list = repo_list
+        self.dest_version = dest_version
+
+    def exec_commaned(self, cmd, expect_failure=False, ignore_failure=False):
+        '''
+        execute command on the appliance and logging the command with it output to log file
+        :param cmd: command
+        :param expect_failure: is the command have to fail
+        :param ignore_failure: not raise exception on command failure
+        :return: command results
+        '''
+
+        logger.debug("exec_commaned: {cmd}".format(cmd=cmd))
+        result = self.ssh_client.run_command(cmd)
+        try:
+            if bool(result.rc) != expect_failure and not ignore_failure:
+                exception_msg = "command {cmd} failed!!\n\n\n" \
+                                "{command_output}".format(cmd=cmd,
+                                                          command_output=result.output)
+                raise RuntimeError(exception_msg)
+        except Exception as ex:
+            logger.exception(ex.message)
+
+        logger.debug(result.output)
+        return result
+
+    def add_yum_repo(self):
+        '''
+        setting all yumrepos before starting the ugrade
+        :return: None
+        '''
+
+        logger.info("adding repositories to yum")
+        for curr_repo_url in self.repo_list:
+            # Add all the repos to yum repo
+            self.exec_commaned("yum-config-manager "
+                               "--add-repo {repo_url}".format(repo_url=curr_repo_url))
+
+    def update_yum(self):
+        '''
+        initiate upgrading process
+        :return: None
+        '''
+        def monitor_yum():
+            '''
+            give real time monitoring on yum update (give the user experience
+            that the script is still running)
+            :return: None
+            '''
+
+            time.sleep(5)
+            # do as long yum is running on the appliance
+            res = self.exec_commaned(""" ps -ef | grep "yum update" | grep -v grep """,
+                                     ignore_failure=True)
+            while not bool(res.rc):
+                logger.info("yum still running")
+                time.sleep(10)
+                res = self.exec_commaned(""" ps -ef | grep "yum update" | grep -v grep """,
+                                         ignore_failure=True)
+
+        # start thr monitoring thread
+        monitor = threading.Thread(target=monitor_yum)
```

Comment:
You are sure you need threads? I found that it is good to try to avoid them as they are rising the code complexity so much. Shouldn't you set `daemon=True`? And/or shouldn't you join the thread at some point? Or you just start start `yum` and just after that restart and have hopeful fingers crossed your binaries got updated before the service restarted?

Threads are advanced stuff. Really.